src/main/kotlin/com/exactpro/th2/conn/dirty/fix/MessageTransformer.kt

```diff
@@ -124,51 +137,111 @@ object MessageTransformer {
+        check(delimiter !in tags) { "'tags' cannot contain 'delimiter' tag: $delimiter" }
+    }
+}
+
+data class Context(val groups: Map<String, Group> = mapOf())
+
+@JsonTypeInfo(use = DEDUCTION)
+@JsonSubTypes(Type(FieldSelector::class), Type(GroupSelector::class))
+interface Selector {
+    fun init(context: Context) = Unit
+    fun find(message: ByteBuf): FixElement?
+    fun find(fields: Iterable<FixField>): FixElement?
+}
+
+fun ByteBuf.contains(selector: Selector): Boolean = selector.find(this) != null
+fun Iterable<FixField>.contains(selector: Selector): Boolean = selector.find(this) != null
+
 data class FieldSelector(
     val tag: Tag?,
-    val tagOneOf: List<Tag>?,
-    val matches: Pattern,
-) {
+    @JsonAlias("one-of-tags", "tagOneOf") val tags: List<Tag>?,
+    @JsonAlias("matching") val matches: Pattern,
+) : Selector {
     init {
-        require((tag != null) xor !tagOneOf.isNullOrEmpty()) { "Either 'tag' or 'tagOneOf' must be specified" }
+        require((tag != null) xor !tags.isNullOrEmpty()) { "Either 'tag' or 'one-of-tags' must be specified" }
     }
 
     @JsonIgnore private val predicate = matches.asMatchPredicate()
 
-    fun matches(message: ByteBuf): Boolean = find(message) != null
-
-    fun find(message: ByteBuf): FixField? = when {
-        tag != null -> message.findField { it.tag == tag && it.value?.run(predicate::test) ?: false }
-        else -> message.findFields { it.tag in tagOneOf!! && it.value?.run(predicate::test) ?: false }.randomOrNull()
+    override fun find(fields: Iterable<FixField>): FixField? = when (tags) {
+        null -> fields.find { it.tag == tag && it.value?.run(predicate::test) ?: false }
+        else -> fields.filter { it.tag in tags && it.value?.run(predicate::test) ?: false }.randomOrNull()
     }
 
+    override fun find(message: ByteBuf): FixField? = find(message.fields)
+
     override fun toString() = buildString {
         tag?.apply { append("tag $tag") }
-        tagOneOf?.apply { append("one of tags $tagOneOf") }
+        tags?.apply { append("one of tags $tags") }
         append(" ~= /$matches/")
     }
 }
 
+data class GroupSelector(
+    val group: String,
+    @JsonAlias("contains", "where") val selectors: List<FieldSelector>,
+) : Selector {
+    private var counter by notNull<Int>()
+    private var delimiter by notNull<Int>()
+    private lateinit var tags: Iterable<Int>
+
+    init {
+        require(selectors.isNotEmpty()) { "group selector has no fields: $group" }
+    }
+
+    override fun init(context: Context) {
+        val group = context.groups[group] ?: error("Unknown group: $group")
+        counter = group.counter
+        delimiter = group.delimiter
+        tags = group.tags
+    }
+
+    override fun find(fields: Iterable<FixField>): FixGroup? {
+        var group = fields.findGroup(counter, delimiter, tags)
+
+        while (group != null) {
+            if (selectors.all(group::contains)) return group
```

Comment:
Could you please add information into README about behavior of this function in case where several groups contains the same values.